Future values of types (#4)
EnergyModelsX/EnergyModelsRecedingHorizon.jl: docs/src/manual/future_value.md

# [Implementation of future values](@id man-fv)

A new [`AbstractElement`](@extref EnergyModelsBase.AbstractElement), [`FutureValue`](@ref), has been implemented to be able to include additional terms to the objective without adding new nodes to the model.
Within `EnergyModelsRecedingHorizon`, we implemented two composite subtypes:

1. The composite type [`StorageValueCuts`](@ref) can be used to describe the value of storages at the end of the optimization period as multiple cutting hyperplanes that depend on one or several of the outgoing state variables.

>    Its implementations is explained below in *[Storage end values](@ref man-fv-sev)*

Reviewer: ```suggestion
   Its implementation is explained below in *[Storage end values](@ref man-fv-sev)*.
```

   For more details about the generation of cuts in stochastic dual dynamic programming we refer to [Dowson2020](@cite).
2. The composite type [`TypeFutureValue`](@ref) adds a value to a given variable for all instances of a specified node.
   It is hence a simplified approach for elements which do not impact each other, *e.g.*, the state of a process.

>    Its implementations is explained below in *[Type end values](@ref man-fv-tev)*

Reviewer: ```suggestion
   Its implementation is explained below in *[Type end values](@ref man-fv-tev)*.
```

    time\_weight(v_{up}) = 1 - time\_weight(v_{down})
\end{aligned}
```

## [Type end values](@id man-fv-tev)

### [Introduced type and its fields](@id man-fv-tev-fields)

The composite type [`TypeFutureValue`](@ref) can be used if you plan to use a single value for a variable for all instances of a node.
In this case, the value is the same.

The fields of a [`TypeFutureValue`](@ref) are given as:

- **`element_type::Type{<:AbstractElement}`**:\
  The node type corresponds to the type for which a variable is resulting in an additional contribution to the cost function.
  The contribution is added for **all** instances of a given type.
- **`val_dict::Dict{Symbol, Real}`**:\
  The value dictionary is including the variables that should have a future value as keys and the corresponding future value as value.
  A positive value corresponds to a beneficial contribution.\
  It is also possible to provide a single variable and value through the implementation of a constructor.

!!! warning "Other AbstractElements"

>     While the functionality is in theory able to be used for any other `AbstractElement`, wecurrently limit it explicitly to nodes to avoid potential method ambiguities in the function `constraints_couple`.

Reviewer: ```suggestion
    While the functionality is in theory able to be used for any other `AbstractElement`, we currently limit it explicitly to nodes to avoid potential method ambiguities in the function `constraints_couple`.
```

### [Mathematical description](@id man-fv-tev-math)

> Given the set of nodes corresponding to the type `element_type` is given ``N^{sub}``, we can calculate the future value as

Reviewer: Just to make it consistent with the above math formulas:
```suggestion
Given the set of nodes corresponding to the type `element_type` is given ``N^{sub}``, we can calculate the future value as:
```

  The contribution is added for **all** instances of a given type.
- **`val_dict::Dict{Symbol, Real}`**:\
  The value dictionary is including the variables that should have a future value as keys and the corresponding future value as value.

>   The value has a corresponding cost value, *e.g.*, €/1000 h when looking at the previous utilization of an electrolyser.

Reviewer: This would be better illustrated by an example.

Author: I agree. Once this PR is registered, I can push the approach in `EnergyModelsHydrogen` and we can revisit the example here. I just did not want to create a node again with a dynamic state to include it. Alternatively, I will add it to one of the examples with the dynamic state, but I do not think it is beneficial for the meaning in these.

  A positive value corresponds to a beneficial contribution.\
  It is also possible to provide a single variable and value through the implementation of a constructor.

!!! warning "Other AbstractElements"
    While the functionality is in theory able to be used for any other `AbstractElement`, we currently limit it explicitly to nodes to avoid potential method ambiguities in the function `constraints_couple`.

!!! danger "Utilizing `TypeFutureValue`"
    `TypeFutureValue` is introduced for a simplified approach for providing a cost contribution to all instances of a given type.

>     It **should** never be used for the reference nodes!

Reviewer: Or maybe node instead of element.
```suggestion
    `TypeFutureValue` is introduced for a simplified approach for providing a cost contribution to all instances of a given element type.
    It should **never** be used for the reference nodes!
```
